Hydroviz hydrographs

Joshdpaul opened the pull request:
This branch does work, but data fetch from Rasdaman is super slow! Like 10 minute return times. Experiments with tiling and splitting into two coverages have helped some, but this really needs to be optimized somehow. Below you can see the tiling scheme chunks all data for 2-10 streams at a time, which I thought would be efficient for querying single streams, but that has proven false.

Also you can see in the dataset printouts during query that the datasets are not very large (<2MB total data) so there is some other bottleneck in the query besides sheer size of the data.

### Coverages & their tiling: 

- `conus_hydro_segments_doy_mmm_maurer_historical_test_2` = 6.94 GB
```
curl -u rasadmin:$PASSWORD   -d 'query=select dbinfo(c,"printtiles=embedded") from conus_hydro_segments_doy_mmm_maurer_historical_test_2 as c'   'https://zeus.snap.uaf.edu/rasdaman/rasql'

{
 "oid": "582145",
 "baseType": "marray <struct { float doy_min, float doy_mean, float doy_max },[D0(*:*),D1(*:*),D2(*:*),D3(*:*),D4(*:*),D5(*:*)]>",
 "setTypeName": "conus_hydro_segments_doy_mmm_maurer_historical_test_2_Set",
 "nullValues": "[nan:nan]",
 "mddTypeName": "conus_hydro_segments_doy_mmm_maurer_historical_test_2_Array",
 "tileNo": "5133",
 "totalSize": "6943224960",
 "tiling": {
	"tilingScheme": "aligned",
	"tileSize": "4194304",
	"tileConfiguration": "[0:*,0:*,0:*,0:*,0:*,0:10]",
	"tileDomains":
	[
		"[0:0,0:365,0:1,0:13,0:0,0:10]",
		"[0:0,0:365,0:1,0:13,0:0,11:21]",
		"[0:0,0:365,0:1,0:13,0:0,22:32]",
		"[0:0,0:365,0:1,0:13,0:0,33:43]",
		"[0:0,0:365,0:1,0:13,0:0,44:54]",
...
                 "[0:0,0:365,0:1,0:13,0:0,46200:46210]",
		"[0:0,0:365,0:1,0:13,0:0,46211:46221]",
		"[0:0,0:365,0:1,0:13,0:0,46222:46232]",
		"[0:0,0:365,0:1,0:13,0:0,46233:46243]",
		"[0:0,0:365,0:1,0:13,0:0,46244:46254]"
	]
 },
 "index": {
	"type": "rpt_index",
	"PCTmax": "4096",
	"PCTmin": "2048"
 }

```


- `conus_hydro_segments_doy_mmm_maurer_projected_test` = 77.37 GB
```
curl -u rasadmin:$PASSWORD   -d 'query=select dbinfo(c,"printtiles=embedded") from conus_hydro_segments_doy_mmm_maurer_projected_test as c'   'https://zeus.snap.uaf.edu/rasdaman/rasql'

{
 "oid": "581633",
 "baseType": "marray <struct { float doy_min, float doy_mean, float doy_max },[D0(*:*),D1(*:*),D2(*:*),D3(*:*),D4(*:*),D5(*:*)]>",
 "setTypeName": "conus_hydro_segments_doy_mmm_maurer_projected_test_Set",
 "nullValues": "[nan:nan]",
 "mddTypeName": "conus_hydro_segments_doy_mmm_maurer_projected_test_Array",
 "tileNo": "18820",
 "totalSize": "77367363840",
 "tiling": {
	"tilingScheme": "aligned",
	"tileSize": "4194304",
	"tileConfiguration": "[0:*,0:*,0:*,0:*,0:*,0:10]",
	"tileDomains":
	[
		"[0:2,0:365,0:1,0:12,0:3,0:2]",
		"[0:2,0:365,0:1,0:12,0:3,3:5]",
		"[0:2,0:365,0:1,0:12,0:3,6:8]",
		"[0:2,0:365,0:1,0:12,0:3,9:11]",
		"[0:2,0:365,0:1,0:12,0:3,12:14]",
...
		"[0:2,0:365,0:1,0:12,0:3,52968:52970]",
		"[0:2,0:365,0:1,0:12,0:3,52971:52973]",
		"[0:2,0:365,0:1,0:12,0:3,52974:52976]",
		"[0:2,0:365,0:1,0:12,0:3,52977:52979]",
		"[0:2,0:365,0:1,0:12,0:3,52980:52982]"
	]
 },
 "index": {
	"type": "rpt_index",
	"PCTmax": "4096",
	"PCTmin": "2048"
 }

```


### Dataset print during query for `http://127.0.0.1:5000/conus_hydrology/hydrograph/1000`:

```
<xarray.Dataset> Size: 127kB
Dimensions:    (era: 1, doy: 366, landcover: 2, model: 14, scenario: 1)
Coordinates:
  * era        (era) <U9 36B '1976-2005'
  * doy        (doy) float64 3kB 1.0 2.0 3.0 4.0 5.0 ... 363.0 364.0 365.0 366.0
  * landcover  (landcover) <U7 56B 'dynamic' 'static'
  * model      (model) <U14 784B 'access1-0' 'bcc-csm1-1' ... 'noresm1-m'
  * scenario   (scenario) <U10 40B 'historical'
Data variables:
    doy_min    (era, doy, landcover, model, scenario) float32 41kB ...
    doy_mean   (era, doy, landcover, model, scenario) float32 41kB ...
    doy_max    (era, doy, landcover, model, scenario) float32 41kB ...
Attributes:
    NCO:                   netCDF Operators version 5.0.6 (Homepage = http://...
    conversion_date:       2025-12-04T16:52:21
    converted_dimensions:  landcover, model, scenario, era
    created:               2025-12-03T15:34:28.121705
    description:           Daily climatology statistics (min, mean, max) by e...
    history:               Fri Dec 12 10:17:25 2025: ncks -O -d scenario,0 hi...
    original_file:         combined_maurer.nc
    rasdaman_conversion:   String dimensions converted to integers for Rasdam...
    title:                 Combined Streamflow Daily Climatologies

<xarray.Dataset> Size: 1MB
Dimensions:    (era: 3, doy: 366, landcover: 2, model: 12, scenario: 4)
Coordinates:
  * era        (era) <U9 108B '2016-2045' '2046-2075' '2071-2100'
  * doy        (doy) float64 3kB 1.0 2.0 3.0 4.0 5.0 ... 363.0 364.0 365.0 366.0
  * landcover  (landcover) <U7 56B 'dynamic' 'static'
  * model      (model) <U14 672B 'access1-0' 'bcc-csm1-1' ... 'mri-cgcm3'
  * scenario   (scenario) <U5 80B 'rcp26' 'rcp45' 'rcp60' 'rcp85'
Data variables:
    doy_min    (era, doy, landcover, model, scenario) float32 422kB ...
    doy_mean   (era, doy, landcover, model, scenario) float32 422kB ...
    doy_max    (era, doy, landcover, model, scenario) float32 422kB ...
Attributes:
    NCO:                   netCDF Operators version 5.0.6 (Homepage = http://...
    conversion_date:       2025-12-04T16:52:21
    converted_dimensions:  landcover, model, scenario, era
    created:               2025-12-03T15:34:28.121705
    description:           Daily climatology statistics (min, mean, max) by e...
    history:               Thu Dec 11 12:42:58 2025: ncks -O -d model,0,7 -d ...
    original_file:         combined_maurer.nc
    rasdaman_conversion:   String dimensions converted to integers for Rasdam...
    title:                 Combined Streamflow Daily Climatologies
```

Joshdpaul commented:
Update:

Tried splitting hydrograph coverages further by landcover and time period:

`conus_hydro_segments_doy_climatology_dynamic_historical` : 3.5GB
`conus_hydro_segments_doy_climatology_static_historical` : 3.5GB
`conus_hydro_segments_doy_climatology_dynamic_projected` : 42GB
`conus_hydro_segments_doy_climatology_static_projected` : 42GB

Requests on the historic coverages take up to 30 seconds, and requests on the projected coverages take up to 2 min. As mentioned in last comment, the size of the data transferred in each query is less than 1MB so the bottleneck here is probably the tiling scheme.

cstephen reviewed (changes requested):
This PR looks/acts great to me! I read everything in the PR description/comments and was bracing for API responses that took up to 3 minutes, but everything is running really fast on my side (as of an ~hour ago). I'm not sure if the subsequent commits improved the performance, but even requests to the `conus_hydrology/hydrograph/<stream segment>` endpoint are returning in less than 3 seconds 😮 

I looked over the output and this does indeed appear to be everything we'll need to implement some compelling looking hydrographs in the webapp, similar to what we did for the Fish & Fire webapp. Also, I think including the metadata in the API response is a smart move that'll help keep the API and webapp in sync during heavy development.

Here are a couple suggestions that came to mind while looking this over:

- Some models are missing RCPs and return empty dictionaries for these RCPs. Using http://localhost:5000/conus_hydrology/hydrograph/1530 as an example, the `access1-0` and `bnu-ems` models are missing some RCPs, and `maurer` is missing all RCPs (as expected since it's the historical model). We generally try to prune the entire key/value pairs for empty dictionaries like this (using the `prune_nulls_with_max_intensity` postprocessing function) for two reasons:
  - It makes the output more succinct and clean.
  - It helps mitigate downstream bugs in webapp code.
- I noticed that the documentation page is currently empty. No biggie here IMO because the dust hasn't settled on this work yet, and things might change. I'd recommend creating a GitHub issue as a reminder that this documentation still needs added.
- I don't have any qualms over changing the `get_axis_encodings` to `get_axis_coordinate_values` in this branch. I've run into similar situations in the past where I need to change some shared code for the purposes of my PR, and this requires slight changes to peripheral files. I've double checked that the `/seaice` endpoints still work as expected in this branch. I think the new version of the `get_axis_encodings` function could benefit from some more comments explaining how it's different from `get_axis_coordinate_values` now, though. It's not super clear to me how it's different by looking at the code, but I assume this code is needed to access other coverage metadata fields, not just axis position information.

Great work, overall!!

Joshdpaul commented:
@cstephen Here is an updated version of this branch with a lot of changes:

- Addressed your most recent comments - I had to make my own pruning function since `prune_nulls_with_max_intensity()` just doesn't work for this output. Maybe because there is a list of dicts in there?
- Merged the branch with USGS gauge endpoint into this one (you had already reviewed that, thanks again!)
- Added CSV generation for all endpoints
- Switched from querying 4 hydrograph coverages to the one big hydrograph coverage, with no discernable drop in performance
- The `http://localhost:5000/conus_hydrology/hydrographs` endpoint has changed to `http://localhost:5000/conus_hydrology/modeled_climatology/` and `http://localhost:5000/conus_hydrology/observed_climatology/` because we aren't actually delivering a hydrograph... this new endpoint name is more descriptive, in my opinion. Dev apps referencing the old endpoint will need to be updated.
- Added the HTML documentation page

To test:

- Proofread the documentation page. Does everything make sense here?
- Check the stats description table against the source table here: https://github.com/ua-snap/hydroviz/blob/main/data/streamflow_statistics_description_table.csv
- Test out the endpoints, including CSV outputs. Try some bad stream IDs (anything above 57,000 or below 0, strings, etc)
- Use the `http://localhost:5000/conus_hydrology/gauge_info` to investigate some streams that have gauges for the `http://localhost:5000/conus_hydrology/observed_climatology/` endpoint. Does this make sense? Is this crossreference useful, and easy enough to find on the documentation page?

cstephen commented on templates/documentation/conus_hydrology.html:
```diff
@@ -0,0 +1,600 @@
+             {"doy": 366, 
+              "doy_max": 12920, 
+              "doy_mean": 4385, 
+              "doy_min": 1706} 
```
The indentation is a little off here. I'd recommend putting each curly brace (`{`, `}`) on its own line and indenting these `doy_*` fields by two more spaces to be consistent with the other output examples.

cstephen commented on csv_functions.py:
```diff
@@ -1307,3 +1316,146 @@ def fire_weather_csv(data, filename_prefix, vars):
+        metadata += "# ma20: Mean of monthly flow values for September.\n"
+        metadata += "# ma21: Mean of monthly flow values for October.\n"
+        metadata += "# ma22: Mean of monthly flow values for November.\n"
+        metadata += "# ma23: Mean of monthly flow values for December.\n"
```
Units for the mean monthly flow can be inferred from other metadata lines, but probably best to add the units for each monthly flow metadata line explicitly (IMO).

cstephen commented on routes/conus_hydrology.py:
```diff
@@ -0,0 +1,672 @@
+                    lat=str(gauge_data_dict["latitude"]),
+                    lon=str(gauge_data_dict["longitude"]),
+                    vars=str(gauge_data_dict["metadata"]["percent_complete"])
+                    + f"% complete",
```
I'd recommend defining a new parameter to use instead of `vars` (both here and in the corresponding `create_csv` function definition in the `csv_functions.py` file). Other API endpoints use the `create_csv` function's `vars` parameter as a filter to determine which variables to include in the CSV, so the way it's being used here is quite a bit different and may lead to trouble in the future.

Joshdpaul replied:
fixed with [44324f2](https://github.com/ua-snap/data-api/pull/674/commits/44324f289e80a02623d8b56aabc9af2f4e6bc98e)

cstephen replied:
Well implemented in 44324f2! Thanks!!

Joshdpaul commented on csv_functions.py:
```diff
@@ -1307,3 +1316,146 @@ def fire_weather_csv(data, filename_prefix, vars):
+        metadata += "# spr_ord: Julian date of spring (April-June) maximum. Determine the Julian date that the maximum flow occurs for each year. SPR_ORD is the median of these values (Julian day - temporal).\n"
+        metadata += "# sum_ord: Julian date of summer (July-September) minimum. Determine the Julian date that the minimum flow occurs for each water year. SUM_ORD is the median of these values (Julian day - temporal).\n"
+        metadata += "# th1: Julian date of annual maximum. Determine the Julian date that the maximum flow occurs for each year. TH1 is the median of these values (Julian day - temporal).\n"
+        metadata += "# tl1: Julian date of annual minimum. Determine the Julian date that the minimum flow occurs for each water year. TL1 is the median of these values (Julian day - temporal).\n"
```
Good question! It would be better to have an additional function that pulls the coverage variable's metadata for populating the CSV metadata. The question is how to pass that info (as a dict, probably) from the route to the csv_functions.create_csv() function. I will ticket this to experiment and see if there is a way to do this.

Joshdpaul commented on templates/documentation/conus_hydrology.html:
```diff
@@ -0,0 +1,600 @@
+    <tr><td><code>spr_ord</code></td><td>Julian date of spring maximum flow.</td><td>Julian day</td></tr>
+    <tr><td><code>sum_ord</code></td><td>Julian date of summer minimum flow.</td><td>Julian day</td></tr>
+    <tr><td><code>th1</code></td><td>Julian date of annual maximum flow.</td><td>Julian day</td></tr>
+    <tr><td><code>tl1</code></td><td>Julian date of annual minimum flow.</td><td>Julian day</td></tr>
```
fixed with [1b2fe1b](https://github.com/ua-snap/data-api/pull/674/commits/1b2fe1bf75a640e54f619eba640d8c29038ecdcb)

cstephen reviewed (approved):
All of my suggestions have been resolved and everything's looking great. Approved! (Not merging until @brucecrevensten  has a chance to look at this.)

Joshdpaul commented on routes/conus_hydrology.py:
```diff
@@ -0,0 +1,685 @@
+            gauge_metadata = await fetch_layer_data(metadata_url, session)
+            gauge_data = await fetch_layer_data(data_url, session)
+    except:
+        return render_template("400/bad_request.html"), 400
```
ticketed #683
